ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/OngoingAuthorizationObserver.scala

```diff
@@ -3,14 +3,84 @@
+import com.daml.lf.data.Ref
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.server.api.validation.ErrorFactories
+import io.grpc.StatusRuntimeException
 import io.grpc.stub.ServerCallStreamObserver
 
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+/** @param userRightsCheckIntervalInSeconds - determines the interval at which to check whether user rights state has changed.
+  *                                          Also, double of this value serves as timeout value for subsequent user rights state checks.
+  */
 private[auth] final class OngoingAuthorizationObserver[A](
     observer: ServerCallStreamObserver[A],
-    claims: ClaimSet.Claims,
-    authorized: ClaimSet.Claims => Either[AuthorizationError, Unit],
-    throwOnFailure: AuthorizationError => Throwable,
-) extends ServerCallStreamObserver[A] {
+    originalClaims: ClaimSet.Claims,
+    nowF: () => Instant,
+    errorFactories: ErrorFactories,
+    userManagementStore: UserManagementStore,
+    userRightsCheckIntervalInSeconds: Int,
+    akkaScheduler: Scheduler,
+)(implicit loggingContext: LoggingContext, ec: ExecutionContext)
+    extends ServerCallStreamObserver[A] {
+
+  private val logger = ContextualizedLogger.get(getClass)
+  private val errorLogger = new DamlContextualizedErrorLogger(logger, loggingContext, None)
+
+  // Guards against propagating calls to delegate observer after either
+  // [[onComplete]] or [[onError]] has already been called once.
+  // We need this because [[onError]] can be invoked two concurrent sources:
+  // 1) scheduled user rights state change task (see [[cancellableO]]),
+  // 2) upstream component that is translating upstream Akka stream into [[onNext]] and other signals.
+  private var afterCompletionOrError = false
+
+  @volatile private var lastUserRightsCheckTime = nowF()
+
+  private lazy val userId = originalClaims.applicationId.fold[Ref.UserId](
+    throw new RuntimeException(
+      "Claims were resolved from a user but userId (applicationId) is missing in the claims."
+    )
+  )(Ref.UserId.assertFromString)
+
+  // Scheduling a task that periodically checks
+  // whether user rights state has changed.
+  // If user rights state has changed it aborts the stream by calling [[onError]]
+  private val cancellableO: Option[Cancellable] = {
+    if (originalClaims.resolvedFromUser) {
+      val delay = userRightsCheckIntervalInSeconds.seconds
+      // Note: https://doc.akka.io/docs/akka/2.6.13/scheduler.html states that:
+      // "All scheduled task will be executed when the ActorSystem is terminated, i.e. the task may execute before its timeout."
+      val c = akkaScheduler.scheduleWithFixedDelay(initialDelay = delay, delay = delay)(runnable =
+        checkUserRights _
+      )
+      Some(c)
+    } else None
+  }
+
+  private def checkUserRights(): Unit = {
+    userManagementStore
+      .listUserRights(userId)
```

Comment:
Interesting point.
We need `userId` only if `if (originalClaims.resolvedFromUser) {` is true so we could restructure the code so that `userId` is resolved after `if (originalClaims.resolvedFromUser) {` but before scheduler starts ticking.

On the other hand, if a stream has started that means it's been authorized so if it was authorized with a user (i.e. `originalClaims.resolvedFromUser` is true) then userId had to be  already successfully resolved somewhere higher up in the authorization stack.